source/common/http/async_client_impl.cc

```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
+    : parent_(parent), stream_callbacks_(callbacks), stream_id_(parent.config_.random_.random()),
+      router_(parent.config_), request_info_(Protocol::Http11),
+      route_(parent_.cluster_.name(), timeout) {
+
+  router_.setDecoderFilterCallbacks(*this);
   // TODO: Correctly set protocol in request info when we support access logging.
 }
 
-AsyncRequestImpl::~AsyncRequestImpl() { ASSERT(!reset_callback_); }
+AsyncStreamImpl::~AsyncStreamImpl() { ASSERT(!reset_callback_); }
 
-void AsyncRequestImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  response_.reset(new ResponseMessageImpl(std::move(headers)));
+void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
 #ifndef NDEBUG
   log_debug("async http request response headers (end_stream={}):", end_stream);
-  response_->headers().iterate([](const HeaderEntry& header, void*) -> void {
+  headers->iterate([](const HeaderEntry& header, void*) -> void {
     log_debug("  '{}':'{}'", header.key().c_str(), header.value().c_str());
   }, nullptr);
 #endif
 
-  if (end_stream) {
-    onComplete();
-  }
+  stream_callbacks_.onHeaders(std::move(headers), end_stream);
+  closeRemote(end_stream);
 }
 
-void AsyncRequestImpl::encodeData(Buffer::Instance& data, bool end_stream) {
+void AsyncStreamImpl::encodeData(Buffer::Instance& data, bool end_stream) {
   log_trace("async http request response data (length={} end_stream={})", data.length(),
             end_stream);
-  if (!response_->body()) {
-    response_->body(Buffer::InstancePtr{new Buffer::OwnedImpl()});
-  }
-  response_->body()->move(data);
-
-  if (end_stream) {
-    onComplete();
-  }
+  stream_callbacks_.onData(data, end_stream);
+  closeRemote(end_stream);
 }
 
-void AsyncRequestImpl::encodeTrailers(HeaderMapPtr&& trailers) {
-  response_->trailers(std::move(trailers));
+void AsyncStreamImpl::encodeTrailers(HeaderMapPtr&& trailers) {
 #ifndef NDEBUG
   log_debug("async http request response trailers:");
-  response_->trailers()->iterate([](const HeaderEntry& header, void*) -> void {
+  trailers->iterate([](const HeaderEntry& header, void*) -> void {
     log_debug("  '{}':'{}'", header.key().c_str(), header.value().c_str());
   }, nullptr);
 #endif
 
-  onComplete();
+  stream_callbacks_.onTrailers(std::move(trailers));
+  closeRemote(true);
 }
 
-void AsyncRequestImpl::cancel() {
-  reset_callback_();
-  cleanup();
+void AsyncStreamImpl::sendHeaders(HeaderMap& headers, bool end_stream) {
+  headers.insertEnvoyInternalRequest().value(Headers::get().EnvoyInternalRequestValues.True);
+  headers.insertForwardedFor().value(parent_.config_.local_info_.address());
+  router_.decodeHeaders(headers, end_stream);
+  closeLocal(end_stream);
 }
 
-void AsyncRequestImpl::onComplete() {
-  complete_ = true;
-  callbacks_.onSuccess(std::move(response_));
-  cleanup();
+void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
+  router_.decodeData(data, end_stream);
+  decoding_buffer_ = &data;
```

Comment:
Sorry just to further clarify, buffering *should* continue to work in the AsyncRequestImpl case (non-streaming).